renku/core/metadata/repository.py

```diff
@@ -0,0 +1,1192 @@
+            if checksum is None:
+                assert revision is not None, "Either ``revision`` or ``checksum`` must be passed."
+                relative_path = os.path.relpath(absolute_path, self.path)
+                prefix = "--filters" if apply_filters else "-p"
+                object = [prefix, f"{revision}:{relative_path}"]
+            else:
+                assert revision is None, "Cannot pass both ``revision`` and ``checksum``."
+                object = ["--filters", "--path", absolute_path] if apply_filters else ["-p"]
+                object.append(checksum)
+
+            command.extend(object)
+            try:
+                subprocess.run(
+                    command,
+                    check=True,
+                    stdout=output_file,
+                    stderr=subprocess.DEVNULL,
+                    cwd=self.path,
+                )
+            except subprocess.CalledProcessError:
+                return False
+            else:
+                # NOTE: No need to check the exit code since ``check=True`` takes care of non-zero exit codes
+                return True
+
+        def get_content_from_submodules():
+            for submodule in self.submodules:
+                try:
+                    Path(absolute_path).relative_to(submodule.path)
+                except ValueError:
+                    continue
+                else:
+                    return submodule.copy_content_to_file(
+                        path=absolute_path, checksum=checksum, revision=revision, apply_filters=apply_filters
+                    )
+
+        if output_file is None:
+            with tempfile.NamedTemporaryFile(mode="w+b", delete=False) as output_file:
+                if get_content_helper():
+                    return output_file.name
+        else:
+            if get_content_helper():
+                return output_file.name
+
+        from_submodules = get_content_from_submodules()
+        if from_submodules:
+            return from_submodules
+
+        raise errors.ExportError(f"File not found in the repository: '{revision}/{checksum}:{path}'")
+
+    def get_object_hash(self, path: Union[Path, str], revision: str = None) -> Optional[str]:
+        """Return git hash of an object in a Repo or its submodule.
+
+        NOTE: path must be relative to the repo's root regardless if this function is called from a subdirectory or not.
+        """
+        absolute_path = get_absolute_path(path, self.path)
+        revision = revision or "HEAD"
+
+        def get_object_hash_from_submodules() -> Optional[str]:
+            for submodule in self.submodules:
+                try:
+                    Path(absolute_path).relative_to(submodule.path)
+                except ValueError:
+                    continue
+                else:
+                    return submodule.get_object_hash(path=absolute_path, revision="HEAD")
+
+        try:
+            return self.run_git_command("rev-parse", f"{revision}:{path}")
```

Comment:
I had to also change this in https://github.com/SwissDataScienceCenter/renku-python/pull/2426/files#diff-1c9cff6463dbad55fd4faaa72d921d9edb9c042365e2efa432b533c83d6bdae6R134 so it works without a revision, using `git hash-object`. We could call the `hash_object()` below if no revision is passed.

Reply:
We can use `HEAD` if no revision is passed. Is there a reason for not doing this? The two methods should return the same results if the object is not modified. However, `hash-object` reads the file and calculate the hash while `rev-parse` reads it from the git metadata. Also, if the file is modified, `rev-parse` returns the hash of the original content.

Reply:
HEAD only works for committed files. With the change I mentioned above that we don't do unnecessary commits, when we do a `renku run` we need to create the entities for the activity for files that are maybe not yet committed (as they'll be committed only once the whole command is finished).